utils/staking-miner/src/dry_run.rs

```diff
@@ -0,0 +1,97 @@
+// Copyright 2021 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! The dry-run command.
+
+use crate::{prelude::*, Signer, SharedConfig, DryRunConfig, WsClient, Error, rpc_helpers::*, params};
+use codec::Encode;
+
+/// Forcefully create the snapshot. This can be used to compute the election at anytime.
+fn force_create_snapshot<T: EPM::Config>(ext: &mut Ext) -> Result<(), Error> {
+	ext.execute_with(|| {
+		if <EPM::Snapshot<T>>::exists() {
+			log::info!(target: LOG_TARGET, "snapshot already exists.");
+			Ok(())
+		} else {
+			log::info!(target: LOG_TARGET, "creating a fake snapshot now.");
+			<EPM::Pallet<T>>::create_snapshot().map(|_| ()).map_err(Into::into)
+		}
+	})
+}
+
+/// Helper method to print the encoded size of the snapshot.
+fn measure_snapshot_size<T: EPM::Config>(ext: &mut Ext) {
+	ext.execute_with(|| {
+		log::info!(target: LOG_TARGET, "Metadata: {:?}", <EPM::Pallet<T>>::snapshot_metadata());
+		log::info!(
+			target: LOG_TARGET,
+			"Encoded Length: {:?}",
+			<EPM::Pallet<T>>::snapshot()
+				.expect("snapshot must exist before calling `measure_snapshot_size`")
+				.encode()
+				.len()
+		);
+	})
+}
+
+/// Fir the stake threshold to in order to have at most `count` voters.
+#[allow(unused)]
+fn find_threshold<T: EPM::Config>(ext: &mut Ext, count: usize) {
+	ext.execute_with(|| {
+		let mut voters = <EPM::Pallet<T>>::snapshot()
+			.expect("snapshot must exist before calling `measure_snapshot_size`")
+			.voters;
+		voters.sort_by_key(|v| v.1);
+		println!("smallest allowed voter is {:?}", voters.into_iter().rev().take(count).last());
```

Comment:
Two comments here:

- Handling this with iteration does some extra work that isn't really necessary, since we have the vector handy. (It would be lots of extra work, but `vec::Iter` is a `DoubleEndedIterator`.)
- Using `take` hides the difference between two cases: where we needed to truncate the voter list, and where we didn't. This matters because if we don't actually need to truncate the list, then we can't use the reported score as the current threshold; smaller voters would be allowed.

Assuming we reverse-sort `voters` as suggested above, we could do:

```suggestion
		match voters.get(count) {
			Some(threshold_voter) => println!("smallest allowed voter is {:?}", threshold_voter),
			None => {
				println!("requested truncation to {} voters but had only {}", count, voters.len());
				println!("smallest current voter: {:?}", voters.last());
			}
		}
```